fem/bilinearform_ext.cpp

```diff
@@ -287,6 +289,417 @@ void PABilinearFormExtension::MultTranspose(const Vector &x, Vector &y) const
+   {
+      int_face_restrict_lex = trialFes->GetFaceRestriction(
+                                 ElementDofOrdering::LEXICOGRAPHIC,
+                                 FaceType::Interior);
+      faceIntX.SetSize(int_face_restrict_lex->Height(), Device::GetMemoryType());
+      faceIntY.SetSize(int_face_restrict_lex->Height(), Device::GetMemoryType());
+      faceIntY.UseDevice(true); // ensure 'faceIntY = 0.0' is done on device
+   }
+
+   if (bdr_face_restrict_lex == NULL && a->GetBFBFI()->Size() > 0)
+   {
+      bdr_face_restrict_lex = trialFes->GetFaceRestriction(
+                                 ElementDofOrdering::LEXICOGRAPHIC,
+                                 FaceType::Boundary,
+                                 L2FaceValues::SingleValued);
+      faceBdrX.SetSize(bdr_face_restrict_lex->Height(), Device::GetMemoryType());
+      faceBdrY.SetSize(bdr_face_restrict_lex->Height(), Device::GetMemoryType());
+      faceBdrY.UseDevice(true); // ensure 'faceBoundY = 0.0' is done on device
+   }
+}
+
+void EABilinearFormExtension::Assemble()
+{
+   SetupRestrictionOperators();
+
+   ne = trialFes->GetMesh()->GetNE();
+   elemDofs = trialFes->GetFE(0)->GetDof();
+
+   ea_data.SetSize(ne*elemDofs*elemDofs, Device::GetMemoryType());
+   ea_data.UseDevice(true);
+   ea_data = 0.0;
+
+   Array<BilinearFormIntegrator*> &integrators = *a->GetDBFI();
+   const int integratorCount = integrators.Size();
+   for (int i = 0; i < integratorCount; ++i)
+   {
+      integrators[i]->AssembleEA(*a->FESpace(), ea_data);
+   }
+
+   faceDofs = trialFes->GetTraceElement(0,
+                                        trialFes->GetMesh()->GetFaceBaseGeometry(0))->GetDof();
+
+   Array<BilinearFormIntegrator*> &intFaceIntegrators = *a->GetFBFI();
+   const int intFaceIntegratorCount = intFaceIntegrators.Size();
+   if (intFaceIntegratorCount>0)
+   {
+      nf_int = trialFes->GetNFbyType(FaceType::Interior);
+      ea_data_int.SetSize(2*nf_int*faceDofs*faceDofs, Device::GetMemoryType());
+      ea_data_ext.SetSize(2*nf_int*faceDofs*faceDofs, Device::GetMemoryType());
+      ea_data_int = 0.0;
+      ea_data_ext = 0.0;
+   }
+   for (int i = 0; i < intFaceIntegratorCount; ++i)
+   {
+      intFaceIntegrators[i]->AssembleEAInteriorFaces(*a->FESpace(),
+                                                     ea_data_int,
+                                                     ea_data_ext);
+   }
+
+   Array<BilinearFormIntegrator*> &bdrFaceIntegrators = *a->GetBFBFI();
+   const int boundFaceIntegratorCount = bdrFaceIntegrators.Size();
+   if (boundFaceIntegratorCount>0)
+   {
+      nf_bdr = trialFes->GetNFbyType(FaceType::Boundary);
+      ea_data_bdr.SetSize(nf_bdr*faceDofs*faceDofs, Device::GetMemoryType());
+      ea_data_bdr = 0.0;
+   }
+   for (int i = 0; i < boundFaceIntegratorCount; ++i)
+   {
+      bdrFaceIntegrators[i]->AssembleEABoundaryFaces(*a->FESpace(),ea_data_bdr);
+   }
+}
+
+void EABilinearFormExtension::AssembleDiagonal(Vector &y) const
```

Comment:
Same as `PABilinearFormExtension::AssembleDiagonal`, depending on what we decide, we could reuse it.

Reply:
agreed.